pkg/components/nodejs/fdextractor.js

```diff
@@ -0,0 +1,164 @@
+    socketWrite:   net.Socket.prototype.write,
+  };
+}
+
+const orig = global[STORE];
+net.Server.prototype.emit  = orig.serverEmit;
+net.Socket.prototype.connect = orig.socketConnect;
+net.Socket.prototype.write = orig.socketWrite;
+
+const { AsyncLocalStorage } = require('async_hooks');
+
+const debug_enabled = false;
+
+function crc8(data) {
+	let crc = 0x00;
+	const polynomial = 0x07;
+
+	for (let i = 0; i < data.length; i++) {
+		crc ^= data[i];
+
+		for (let bit = 0; bit < 8; bit++) {
+			if (crc & 0x80) {
+				crc = ((crc << 1) ^ polynomial) & 0xFF;
+			} else {
+				crc = (crc << 1) & 0xFF;
+			}
+		}
+	}
+
+	return crc;
+}
+
+const ipcServer = net.createServer();
+
+ipcServer.listen(0, '127.0.0.1', () => {
+	if (!debug_enabled) return;
+
+	const addr = ipcServer.address();
+	console.log(`[ipc] server listening`);
+});
+
+let ipcClient;
+
+ipcServer.on('listening', () => {
+	const { port, address } = ipcServer.address();
+	ipcClient = net.connect(port, address, () => {
+		ipcClient.setNoDelay(true)
+		if (debug_enabled)
+			console.log(`[ipc] client connected to server`);
+	});
+});
+
+ipcServer.on('connection', socket => {
+	socket.on('data', data => {
+		if (!debug_enabled || data.length < 20) return;
+		const marker = data.readUInt32BE(0);
+		const evType = data.readUInt8(4);
+		const len = data.readUInt8(5);
+		const inFd   = data.readUInt32BE(8);
+		const outFd  = data.readUInt32BE(12);
+		console.log(
+			`[ipc] marker=0x${marker.toString(16)}, t=${evType} len=${len} inFd=${inFd}, outFd=${outFd}`
+		);
+	});
+});
+
+// ebpf ipc marker constant (32-bit)
+const MARKER = 0xBE14BE14;
+
+// ALS store holds only incomingFd
+const als = new AsyncLocalStorage();
+
+net.Server.prototype.emit = function(event, ...args) {
+	if (event === 'connection') {
+		const socket = args[0];
+		const incomingFd = socket._handle && socket._handle.fd;
+
+		if (debug_enabled) {
+			console.log(
+				`[incoming TCP] fd:${incomingFd}, remote=${socket.remoteAddress}:${socket.remotePort}`
+			);
+		}
+
+		return als.run({ incomingFd }, () =>
+			orig.serverEmit.call(this, event, ...args)
+		);
+	}
+	return orig.serverEmit.call(this, event, ...args);
+};
+
+function correlate(incomingFd, outFd, socket) {
+	if (incomingFd < 0 || outFd < 0 || incomingFd === outFd) {
+		return Promise.resolve();
+	}
+
+	const addr = socket.remoteAddress || 'unknown';
+	const port = socket.remotePort || 'unknown';
+
+	if (debug_enabled) {
+		console.log(
+			`[outgoing TCP] inFd:${incomingFd}, outFd:${outFd}, to=${addr}:${port}`
+		);
+	}
+
+	const nodeJSEventType = 0;
+	const buf = Buffer.alloc(20);
+	buf.writeUInt32BE(MARKER, 0);
+	buf.writeUInt8(nodeJSEventType, 4);
+	buf.writeUInt8(buf.length, 5);
+	buf.writeUInt8(0, 6);
+	buf.writeUInt8(0, 7);
+	buf.writeUInt32BE(incomingFd, 8);
+	buf.writeUInt32BE(outFd, 12);
+	buf.writeUInt32BE(0, 16);
+
+	const crc = crc8(buf.slice(0, -1))
+	buf.writeUInt8(crc, 19);
+
+	if (ipcClient && ipcClient.writable) {
```

Comment:
Nit: There are some formatting issues, it might be good to run some sort of js formatter on this.